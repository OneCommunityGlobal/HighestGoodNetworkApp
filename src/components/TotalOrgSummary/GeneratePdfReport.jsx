// generatePDF.js
import jsPDF from 'jspdf';
import html2canvas from 'html2canvas';
import './TotalOrgSummary.css';

export function generatePDF() {
  const input = document.getElementById('pdfContent');
  
  // 添加 pdf-mode 类使打印样式生效
  input.classList.add('pdf-mode');

  // 延时确保样式生效后再截图
  setTimeout(() => {
    // 等待所有字体加载完成
    document.fonts.ready.then(() => {
      html2canvas(input, {
        scale: window.devicePixelRatio || 2,
        useCORS: true,
        allowTaint: false,
        scrollY: -window.scrollY
      }).then((canvas) => {
        // 截图完成后移除 pdf-mode 类
        input.classList.remove('pdf-mode');

        const imgData = canvas.toDataURL('image/png');
        const pdf = new jsPDF('p', 'mm', 'a4');
        const pageWidth = pdf.internal.pageSize.getWidth();
        const pageHeight = pdf.internal.pageSize.getHeight();

        // 计算图片在PDF中的显示尺寸
        const imgWidth = pageWidth;
        const imgHeight = (canvas.height * imgWidth) / canvas.width;
        let position = 0;

        pdf.addImage(imgData, 'PNG', 0, position, imgWidth, imgHeight);

        // 如果内容超过一页，处理分页
        let heightLeft = imgHeight;
        while (heightLeft > pageHeight) {
          position = heightLeft - imgHeight;
          pdf.addPage();
          pdf.addImage(imgData, 'PNG', 0, -position, imgWidth, imgHeight);
          heightLeft -= pageHeight;
        }
        
        pdf.save('document.pdf');
      });
    });
  }, 100); // 延时100ms，视情况可调整
}