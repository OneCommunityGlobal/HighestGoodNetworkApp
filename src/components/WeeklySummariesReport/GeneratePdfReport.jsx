import PropTypes from 'prop-types';
import React from 'react';
import pdfMake from 'pdfmake/build/pdfmake';
import pdfFonts from 'pdfmake/build/vfs_fonts';
import htmlToPdfmake from 'html-to-pdfmake';
import moment from 'moment-timezone';
import { Button } from 'reactstrap';

pdfMake.vfs = pdfFonts.pdfMake.vfs;

const GeneratePdfReport = ({ summaries, weekIndex, weekDates }) => {

  const generateFormattedReport = () => {

    const emails = [];

    let wsReport = `<h1>Weekly Summaries Report</h1>
    <div style="margin-bottom: 20px; color: orange;">From <b>${weekDates.fromDate}</b> to <b>${weekDates.toDate}</b></div>`;
    
    const weeklySummaryNotProvidedMessage = '<div><b>Weekly Summary:</b> <span style="color: red;">Not provided!</span></div>';

    summaries.sort((a, b) => (`${a.firstName} ${a.lastName}`).localeCompare(`${b.firstName} ${b.lastname}`));

    summaries.forEach((eachSummary) => {

      if(eachSummary.email !== undefined && eachSummary.email !== null) {
        emails.push(eachSummary.email);
      }

      const {
        firstName, lastName, weeklySummaries,
        mediaUrl, weeklySummariesCount, weeklyComittedHours
      } = eachSummary;
  
      const mediaUrlLink = mediaUrl ? `<a href=${mediaUrl}>Open link to media files</a>` : '<span style="color: red;">Not provided!</span>';
      const totalValidWeeklySummaries = weeklySummariesCount || 'No valid submissions yet!';
      let weeklySummaryMessage = weeklySummaryNotProvidedMessage;
      if (Array.isArray(weeklySummaries) && weeklySummaries.length && weeklySummaries[weekIndex]) {
        const { dueDate, summary } = weeklySummaries[weekIndex];
        if (summary) {
          weeklySummaryMessage = `<div><b>Weekly Summary</b> (for the week ending on <b>${moment(dueDate).tz('America/Los_Angeles').format('YYYY-MMM-DD')}</b>):</div>
                                  <div data-pdfmake="{&quot;margin&quot;:[20,0,20,0]}">${summary}</div>`;
        }
      }
  
      wsReport += `\n
      <div><b>Name:</b> <span class="name">${firstName} ${lastName}</span></div>
      <div><b>Media URL:</b> ${mediaUrlLink}</div>
      <div style="${totalValidWeeklySummaries === 8 ? 'text-decoration: underline; color: red;' : ''}"><b>Total valid weekly summaries:</b> ${totalValidWeeklySummaries}</div>
      <div><b>Committed weekly hours:</b> ${weeklyComittedHours}</div>
      ${weeklySummaryMessage}
      <div style="color:#DEE2E6; margin:10px 0px 20px 0px; text-align:center;">_______________________________________________________________________________________________</div>`;
    });

    wsReport += '<h2>Emails</h2>'

    wsReport += [...(new Set(emails))] //elimiates duplicate entries if they're somehow present
    .toString()
    .slice(1, emails.toString().length - 1)
    .replaceAll(',', ', ');
    

    return wsReport;
  };
  
  const generateAndDownloadPdf = () => {
  
    const html = htmlToPdfmake(generateFormattedReport());
  
    const docDefinition = {
      content: [
        html,
      ],
      styles: {
        'html-div': { margin: [0, 4, 0, 4] },
        name: {
          background: 'yellow',
        },
      },
    };
  
    pdfMake.createPdf(docDefinition).download(`WeeklySummary-${weekDates.fromDate}-to-${weekDates.toDate}`);
  
  };

  return (
    <Button className="px-5 btn--dark-sea-green float-right" onClick={generateAndDownloadPdf}>Open PDF</Button>
  );
};

GeneratePdfReport.propTypes = {
  summaries: PropTypes.arrayOf(PropTypes.object).isRequired,
  weekDates: PropTypes.shape({
    fromDate: PropTypes.string,
    toDate: PropTypes.string,
  }).isRequired,
  weekIndex: PropTypes.string.isRequired,
};

export default GeneratePdfReport;
