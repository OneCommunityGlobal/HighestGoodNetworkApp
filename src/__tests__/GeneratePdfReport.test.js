import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import mockSummaries from 'weeklySummariesReportData'; // Located in the tested component's __mocks__ folder
import GeneratePdfReport from '../components/WeeklySummariesReport/GeneratePdfReport';

const getWeekDates = () => ({
  fromDate: 'June 14th',
  toDate: 'June 20th, 2020',
});

describe('FormattedReport Component', () => {
  it('Snapshot with mocked data', () => {
    const { asFragment } = render(
      <GeneratePdfReport summaries={mockSummaries} weekIndex="1" weekDates={getWeekDates()} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
