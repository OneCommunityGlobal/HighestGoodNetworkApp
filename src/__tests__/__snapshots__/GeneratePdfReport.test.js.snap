// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedReport Component Snapshot with mocked data 1`] = `
<DocumentFragment>
  <button
    class="btn--dark-sea-green btn btn-secondary"
    style="box-shadow: 2px 2px 4px 1px lightgray;"
    type="button"
  >
    Open PDF
  </button>
</DocumentFragment>
`;
